registry/registry.go

```diff
@@ -74,79 +76,19 @@ func (f roundtripperFunc) RoundTrip(r *http.Request) (*http.Response, error) {
 //   foo/helloworld         -> index.docker.io/foo/helloworld
 //   quay.io/foo/helloworld -> quay.io/foo/helloworld
 //
-func (c *client) GetRepository(repository string) (_ []flux.ImageDescription, err error) {
-	defer func(start time.Time) {
-		c.Metrics.FetchDuration.With(
-			LabelRepository, repository,
-			fluxmetrics.LabelSuccess, strconv.FormatBool(err == nil),
-		).Observe(time.Since(start).Seconds())
-	}(time.Now())
-
-	var host, org, image string
-	parts := strings.Split(repository, "/")
-	switch len(parts) {
-	case 1:
-		host = dockerHubHost
-		org = dockerHubLibrary
-		image = parts[0]
-	case 2:
-		host = dockerHubHost
-		org = parts[0]
-		image = parts[1]
-	case 3:
-		host = parts[0]
-		org = parts[1]
-		image = parts[2]
-	default:
-		return nil, fmt.Errorf(`expected image name as either "<host>/<org>/<image>", "<org>/<image>", or "<image>"`)
+func (c *registry) GetRepository(img Image) (_ []Image, err error) {
+	var remote Remote
+	{
+		remote, err = c.factory.Create(img)
+		remote = NewRemoteMonitoringMiddleware(c.Metrics)(remote)
```

Comment:
A simpler way? This is the bestest way evaa! Middlewares FTW. It means that the remote doesn't need to know anything about monitoring. Decoupled and cohesive.